pkg/apis/unikorn/v1alpha1/types.go

```diff
@@ -517,3 +517,129 @@ type FlavorQuota struct {
 
 type QuotaStatus struct {
 }
+
+// SecurityGroupList is a typed list of security groups.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type SecurityGroupList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []SecurityGroup `json:"items"`
+}
+
+// SecurityGroup defines a security group beloning to an identity.
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced,categories=unikorn
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="status",type="string",JSONPath=".status.conditions[?(@.type==\"Available\")].reason"
+// +kubebuilder:printcolumn:name="age",type="date",JSONPath=".metadata.creationTimestamp"
+type SecurityGroup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              SecurityGroupSpec   `json:"spec"`
+	Status            SecurityGroupStatus `json:"status,omitempty"`
+}
+
+type SecurityGroupSpec struct {
+	// Pause, if true, will inhibit reconciliation.
+	Pause bool `json:"pause,omitempty"`
+	// Tags are an abitrary list of key/value pairs that a client
+	// may populate to store metadata for the resource.
+	Tags TagList `json:"tags,omitempty"`
+	// Provider defines the provider type.
+	Provider Provider `json:"provider"`
+	// Ingress are the ingress rules.
+	Ingress []SecurityGroupRule `json:"ingress,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=tcp;udp
+type SecurityGroupRuleProtocol string
+
+const (
+	TCP SecurityGroupRuleProtocol = "tcp"
+	UDP SecurityGroupRuleProtocol = "udp"
+)
+
+type SecurityGroupRulePortRange struct {
+	// Start is the start of the range.
+	// +kubebuilder:validation:Minimum=1
+	Start int `json:"start"`
+	// End is the end of the range.
+	// +kubebuilder:validation:Maximum=65535
+	End int `json:"end"`
+}
+
+type SecurityGroupRulePort struct {
+	// Number is the port number.
+	Number *int `json:"number,omitempty"`
+	// Range is the port range.
+	Range *SecurityGroupRulePortRange `json:"range,omitempty"`
+}
+
+type SecurityGroupRule struct {
```

Comment:
I think the best thing we can do here for general cleanliness and ease of use is to have each rule have a unique identifier and also a description e.g.

```yaml
kind: SecurityGroup
spec:
  ingress:
  - id: 9575b2ef-2eee-4c75-a6e8-59537984185a
    description: OpenSSH
    port:
      number: 22
```

The description is just to make the UX a lot better, sure Adam will thank you in future!

The ID however can be used in the mapping from unikorn to openstack e.g.

```yaml
kind: OpenstackSecurityGroup
spec:
  id: 13a4f72a-9467-4262-be73-021d44541cc8
  rules:
  - id: 9575b2ef-2eee-4c75-a6e8-59537984185a
    openstackID: f7fc698a-628a-4652-8355-9d3b69c9b120
```

I think this will improve the reconcile logic a whole heap as you're just concerned with IDs and not the whole tuple of direction, port range, protocol etc etc.  Obviously the API will need to be tweaked to be:

```
GET/POST /.../securitygroups
GET?PUT/DELETE /.../securitygroups/{id}
GET/POST /.../securitygroups/{id}/rules
GET/PUT/DELETE /.../securitygroups/{id}/rules/{id}
```

A call to create a rule will generate a unique UUID like creation of the security group itself.  I am aware of time, so rather than have rules as a separate CRD and controller we leave that bit the way it is.

As I always say, if the data structure is not quite right, the code will suck!